### PR TITLE
Allow connecting signals to existing methods without opening the script editor

### LINKED ISSUE #8351: Add an editor setting to disable switching to script editor after connecting a signal
### Describe the project you are working on

A 2D top-down colony survival game set in space

### Describe the problem or limitation you are having in your project

I'm programming my game in C# in VSCode.

When it comes to signals, my programming workflow is as follows:
1. In VSCode, create methods that will be triggered by signals (not just stubs but fully functional methods)
2. In Godot, double click a signal to open the signal connection dialog
3. Select the receiving node and pick the matching method using the method list window
4. Click the `Connect` button

Since my method is already implemented and I don't need to change it further, I don't need to switch to VSCode after connecting the signal.
However, clicking `Connect` in 4.1 will always switch me back to VSCode. I want to avoid that.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I want to add a new editor setting to disable the automatic switch to the text editor.
Using this setting I won't have to switch back to Godot every time I connect a signal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add a new editor setting to disable the automatic switch to the text editor when connecting a signal.

Ideally, you could also choose what to do directly in the signal connection dialog with a toggle. The toggle default would be equal to the editor setting value. I don't think this modification is strictly necessary though.

After looking through the existing categories, I think adding the new setting under `interface/editor` is the most appropriate (it kinda feel like a misc category right now though).

The setting itself would be called `Open text editor on signal connection` and would be true by default (to preserve the current behavior).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't know of any way to work around this behavior.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't think it's doable without changing the editor code itself.

### PR DESCRIPTION
**Pull request by passivestar**
Closes https://github.com/godotengine/godot-proposals/issues/8351

Adds an editor setting to prevent the script editor from opening when connecting a signal to an existing method

Screenshot 2024-03-08 at 03 40 27

Note that script editor opening when connecting to an existing method [was a bug](https://github.com/godotengine/godot-proposals/issues/8351#issuecomment-1893996727), but we're preserving that behavior by default

_EDIT: Switched from doing this by holding Ctrl when pressing Connect to an editor setting_

### CODE REVIEW

**Comment by Mickeon** [comment]
While I do find this to be a decent feature, the editor has a discoverability issue for these kinds of things and this additional CTRL shortcut does not help.

At least, if possible, what I would do is have the text on the change **completely** when holding the CTRL key. I've never seen something like this done by the editor before, but _at least_ it makes the feature more noticeable.

In my opinion though, this could just as easily be an Editor Setting, or perhaps, a toggle on the Connect dialog itself, or both?

File: scene/gui/dialogs.cpp [hunk 1]
```diff
@@ -195,6 +195,14 @@ String AcceptDialog::get_ok_button_text() const {
...
 }
 
+void AcceptDialog::set_ok_button_tooltip_text(String p_ok_button_tooltip_text) {
+	ok_button->set_tooltip_text(p_ok_button_tooltip_text);
+}
+
+String AcceptDialog::get_ok_button_tooltip_text() const {
+	return ok_button->get_tooltip_text();
+}
+
```
Comment (Mickeon) [inline scene/gui/dialogs.cpp:205]: Raising eyebrows, these new methods feel a little _too_ niche. Should not exist unless other things in the engine would greatly benefit from them.

File: editor/connections_dialog.cpp [hunk 2]
```diff
@@ -861,6 +861,10 @@ ConnectDialog::ConnectDialog() {
 	error->set_title(TTR("Cannot connect signal"));
 	error->set_ok_button_text(TTR("Close"));
 	set_ok_button_text(TTR("Connect"));
+
+	bool macos = OS::get_singleton()->has_feature("macos") || OS::get_singleton()->has_feature("web_macos") || OS::get_singleton()->has_feature("web_ios");
+	String keycode_name = macos ? find_keycode_name(Key::META) : find_keycode_name(Key::CTRL);
```
[the remark this answers is not in the document]
Comment (passivestar) [reply editor/connections_dialog.cpp:866]: true

File: scene/gui/dialogs.cpp (hunk 1 above)
Comment (passivestar) [reply scene/gui/dialogs.cpp:205]: > get_ok_button() already exists, so these methods are not needed.

ah nice

**Comment by passivestar** [comment]
> At least, if possible, what I would do is have the text on the change completely when holding the CTRL key. I've never seen something like this done by the editor before, but at least it makes the feature more noticeable.

I've never seen anything like that either and I'm not sure it would help discoverability if you need to hold CTRL to discover it

> In my opinion though, this could just as easily be an Editor Setting, or perhaps, a toggle on the Connect dialog itself, or both?

I'm 100% on board to make it an editor setting (with default being the current behavior) and nothing else. In my opinion it's all that's really needed here, especially considering that editor opening on existing methods was unintended

**Comment by KoBeWi** [comment]
Seems like it breaks connecting to native methods:

https://github.com/godotengine/godot/assets/2223172/35b9937e-c4d5-4d4a-8aa1-d7e5e77016b3

I held Ctrl, but not only it opened the editor, it redefined `queue_free()`.

**Comment by passivestar** [comment]
> Seems like it breaks connecting to native methods

Fixed that and also switched from the modifier key to an editor setting

Here's how it works now:

1. If the method exists in an inherited class, either native or script, it creates a connection without adding anything or opening the editor (as it did before)
2. If the method exists in the script on the target node itself it checks the editor setting to decide whether to open the editor or not

I put it into **"Text Editor -> Behavior -> Navigation"** (let me know if there's a better place)

Screenshot 2024-03-08 at 03 40 27

File: doc/classes/EditorSettings.xml [hunk 3]
```diff
@@ -1011,6 +1011,9 @@
 		<member name="text_editor/behavior/navigation/move_caret_on_right_click" type="bool" setter="" getter="">
 			If [code]true[/code], the caret will be moved when right-clicking somewhere in the script editor (like when left-clicking or middle-clicking). If [code]false[/code], the caret will only be moved when left-clicking or middle-clicking somewhere.
 		</member>
+		<member name="text_editor/behavior/navigation/open_script_when_connecting_signal_to_existing_method" type="bool" setter="" getter="">
+			If [code]true[/code], opens the script editor when connecting a signal to an existing method from the Node dock.
```
Comment (KoBeWi) [inline doc/classes/EditorSettings.xml:1015]:
```suggestion
			If [code]true[/code], opens the script editor when connecting a signal to an existing script method from the Node dock.
```
More precise.

File: editor/connections_dialog.cpp [hunk 4]
```diff
@@ -906,25 +906,34 @@ void ConnectionsDock::_make_or_edit_connection() {
...
-		Ref<Script> inherited_scr = scr->get_base_script();
-		while (!inherited_scr.is_null()) {
-			int line = inherited_scr->get_language()->find_function(cd.method, inherited_scr->get_source_code());
-			if (line != -1) {
-				found_inherited_function = true;
-				break;
+		// Check in target's own script.
+		int line = scr->get_language()->find_function(cd.method, scr->get_source_code());
+		if (line != -1) {
+			add_script_function_request = bool(EDITOR_GET("text_editor/behavior/navigation/open_script_when_connecting_signal_to_existing_method"));
```
Comment (KoBeWi) [inline editor/connections_dialog.cpp:920]:
```suggestion
			add_script_function_request = EDITOR_GET("text_editor/behavior/navigation/open_script_when_connecting_signal_to_existing_method");
```
No need to be explicit when assigning.

**Review by KoBeWi — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**